### ISSUE TITLE
No sample for DownloadDirectoryAsync()

### ISSUE THREAD

**Issue by BuggyLogic** [Azure/azure-storage-net-data-movement#54]
Any chance you could add a sample for how to use TransferManager.DownloadDirectoryAsync(). That is a scenario you are currently not covering (you have an example for upload, but not download). I'm having some difficulties, and an example would be most welcome.

**Comment by saschagottfried** [comment]
Since I am having difficulties as well - transfer state shows neither transferred nor skipped file or failure

```
Data movement directory download sample.
Final transfer state: Transferred bytes: 0; Transfered: 0; Skipped: 0, Failed: 0
Files in directory http://127.0.0.1:10000/devstoreaccount1/samplecontainer/blobdir/ downloading to . is finished.

```

What I have done? I just extended the [DataMovementSamples ](https://github.com/Azure/azure-storage-net-data-movement/blob/master/samples/DataMovementSamples/DataMovementSamples/Samples.cs)- since there was no sample covering `DownloadDirectoryAsync`

```
 private static async Task BlobDirectoryDownloadSample()
        {
            string destDirPath = ".";
            CloudBlobDirectory sourceDir = await Util.GetCloudBlobDirectoryAsync(ContainerName, "blobdir");

            // SearchPattern and Recuresive can be used to determine the files to be transferred from the source directory. The behavior of
            // SearchPattern and Recuresive varies for different source directory types.
            // See https://azure.github.io/azure-storage-net-data-movement for more details.
            //
            // When source is local directory, data movement library matches source files against the SearchPattern as standard wildcards. If 
            // recuresive is set to false, only files directly under the source directory will be matched. Otherwise, all files in the
            // sub-directory will be matched as well.
            //
            // In the following case, data movement library will upload png files starting with "azure" in the source directory as block blobs,
            // not including the sub-directory.
            DownloadDirectoryOptions options = new DownloadDirectoryOptions()
            {
                SearchPattern = "*.*",
                Recursive = true             
            };

            // Register for transfer event.
            DirectoryTransferContext context = new DirectoryTransferContext();            

            // Start the upload
            var transferStatus = await TransferManager.DownloadDirectoryAsync(sourceDir, destDirPath, options, context);

            Console.WriteLine("Final transfer state: {0}", TransferStatusToString(transferStatus));
            Console.WriteLine("Files in directory {0} downloading to {1} is finished.", sourceDir.Uri, destDirPath);
        }
```

I looked for a test covering TransferManager.DownloadDirectoryAsync(). But I could not find any. I am still investigating my issue, will come back later on ...

In the meantime any comment from the developers on this could be helpful. Is there anybody out there?

**Comment by EmmaZhu** [comment]
When source is blob directory, SearchPattern means Prefix the blobs you'd want to transfer. In your case, you can just leave SearchPattern not assigned and set Recursive to true to transfer all blobs in the directory. 

I've opened a bug to add samples about directory downloading.

**Comment by saschagottfried** [comment]
@EmmaZhu, please provide a link to the opened bug so that I can subscribe to this issue. Thank you!

**Comment by EmmaZhu** [comment]
@saschagottfried  The bug tracking is internal, you may not have access to it. We'll update this issue thread when the sample is released. 

Did you try with my last answer? Does it help?

**Comment by saschagottfried** [comment]
@EmmaZhu With `searchPattern` assigned no files are matched. I verified that `DownloadDirectoryOptions.SearchPattern" rather works as a prefix. Nevertheless this beahviour is a bit misleading to me (may be to other developers as well), especially when no code samples are provided. 

When not assigned, as you suggested, the code is downloading the blob directory properly. Thank you for helping.

```
DownloadDirectoryOptions options = new DownloadDirectoryOptions
{
	//SearchPattern = "*.*",
	Recursive = true	
};

// Register for transfer event.
DirectoryTransferContext context = new DirectoryTransferContext();

// Start the download
var transferStatus = await TransferManager.DownloadDirectoryAsync(sourceDir, destDirPath, options, context);

Console.WriteLine("Final transfer state: {0}", TransferStatusToString(transferStatus));
Console.WriteLine("Files in directory {0} downloading to {1} is finished.", sourceDir.Uri, destDirPath);
```
Transfer state now displays number of transferred files.
```
Final transfer state: Transferred bytes: 47169094; Transfered: 92; Skipped: 0, Failed: 0
Files in directory http://127.0.0.1:10000/devstoreaccount1/samplecontainer/DML/ downloading to C:\DMLDownloads is finished.
```

**Comment by saschagottfried** [comment]
According to [AzCopy parameters docs](https://docs.microsoft.com/en-us/azure/storage/storage-use-azcopy#azcopy-parameters) 

> /Pattern:"file-pattern
> If the specified source is a blob container or virtual directory, then wildcards are not applied. If option /S is specified, then AzCopy interprets the specified file pattern as a blob prefix. If option /S is not specified, then AzCopy matches the file pattern against exact blob names.

According to AzCopy docs the code below downloads all files beginning with "R". It works as a prefix even when `recursive` option is enabled. Code is working - but differs from what is known to AzCopy users.
```
DownloadDirectoryOptions options = new DownloadDirectoryOptions
{
	SearchPattern = "R",
	Recursive = true                
};
```

According to AzCopy docs the code below should download any files with given prefix. But that does not happen.
```
DownloadDirectoryOptions options = new DownloadDirectoryOptions
{
	SearchPattern = "any-prefix",
	Recursive = false                
};
```

```
Final transfer state: Transferred bytes: 0; Transfered: 0; Skipped: 0, Failed: 0
Files in directory http://127.0.0.1:10000/devstoreaccount1/samplecontainer/DML/ downloading to C:\DMLDownloads is finished.
```

Bottom line: DML is different a bit from AzCopy behaviour. `DownloadDirectoryOptions` has a subtle semantic mismatch. Is any of this a bug?

**Comment by EmmaZhu** [comment]
AzCopy internally uses DMLib to transfer data. Their behaviors on Pattern are the same that wildcards won't work in /Pattern (in AzCopy) or SearchPattern (in DMLib) when source is blob service. Wildcards only work when source is local file folder.  
When source is blob service, the SearchPattern means:
1. Prefix of blob name if Recursive is true
2. The exact match blob name, if Recursive is false.

**Comment by EmmaZhu** [comment]
AzCopy internally uses DMLib to transfer data. Their behaviors on Pattern are the same that wildcards won't work in /Pattern (in AzCopy) or SearchPattern (in DMLib) when source is blob service. Wildcards only work when source is local file folder.
When source is blob service, the SearchPattern means:
Prefix of blob name if Recursive is true
The exact match blob name, if Recursive is false.
Hope this helps.

**Comment by Surfinsanta** [comment]
This was adapted from [https://github.com/Azure/azure-storage-net-data-movement/blob/master/samples/DataMovementSamples/DataMovementSamples/Samples.cs]().

```
public static async Task DownloadDirectoryAsyncExample()
{
    // Build cloud storage account
    string accountName = "Your Account Name";
    string accountKey = "Your Account Key";
    string storageConnectionString = "DefaultEndpointsProtocol=https;AccountName=" + accountName + ";AccountKey=" + accountKey;
    CloudStorageAccount act = CloudStorageAccount.Parse(storageConnectionString);

    // Build cloud blob directory
    CloudBlobClient cbc = act.CreateCloudBlobClient();
    string containerName = "Your Container Name";
    string folderName = "Your Folder Name";
    CloudBlobContainer container = cbc.GetContainerReference(containerName);
    CloudBlobDirectory cbd = container.GetDirectoryReference(folderName);

    // Download directory to desktop
    string path = Environment.GetFolderPath(Environment.SpecialFolder.Desktop) + "\\" + folderName;

    // Get all subdirectories
    DownloadDirectoryOptions ddo = new DownloadDirectoryOptions()
    {
        Recursive = true
    };

    // Create directory transfer context
    TransferCheckpoint tc = null;
    DirectoryTransferContext dtc = GetDirectoryTransferContextExample(tc);

    // Setup task to be monitored
    Stopwatch stopWatch = Stopwatch.StartNew();
    Task task;

    try
    {
        task = TransferManager.DownloadDirectoryAsync(cbd, path, ddo, dtc);
        while (!task.IsCompleted) { /* Hang ten */ }
        await task;
    }
    catch (Exception e)
    {
        Console.WriteLine(e.Message);
    }

    stopWatch.Stop();
    Console.WriteLine("\nTransfer operation completed in " + stopWatch.Elapsed.TotalSeconds + " seconds.");
}
```
```
public static DirectoryTransferContext GetDirectoryTransferContextExample(TransferCheckpoint checkpoint)
{
    DirectoryTransferContext context = new DirectoryTransferContext(checkpoint);

    context.ProgressHandler = new Progress<TransferStatus>((progress) =>
    {
        Console.Write("\rBytes transferred: {0}", progress.BytesTransferred);
    });

    return context;
}
```
